libraries/AP_Telemetry/AP_Telemetry_MQTT.cpp

```diff
@@ -0,0 +1,252 @@
+char const* AP_Telemetry_MQTT::mqtt_server;
+int mqtt_msg_arrived(void *context, char *topicname, int topicLen, MQTTAsync_message* message);
+void onConnect(void *context, MQTTAsync_successData* response);
+void onConnectFailure(void* context, MQTTAsync_failureData* response);
+
+// MQTT Client accessor
+MQTTAsync* AP_Telemetry_MQTT::get_MQTTClient()
+{
+    return &mqtt_client;
+}
+
+AP_Telemetry_MQTT* AP_Telemetry_MQTT::get_telemetry_mqtt()
+{
+    return telemetry_mqtt;
+}
+
+AP_Telemetry_MQTT* AP_Telemetry_MQTT::init_telemetry_mqtt(AP_Telemetry &frontend, AP_HAL::UARTDriver* uart)
+{
+    if (telemetry_mqtt == nullptr) {
+        telemetry_mqtt = new AP_Telemetry_MQTT(frontend, uart);
+        telemetry_mqtt->init_mqtt();
+    }
+    return telemetry_mqtt;
+}
+
+AP_Telemetry_MQTT::AP_Telemetry_MQTT(AP_Telemetry &frontend, AP_HAL::UARTDriver* uart) :
+    AP_Telemetry_Backend(frontend, uart)
+{}
+
+void AP_Telemetry_MQTT::init_mqtt()
+{
+    int rc;
+    mqtt_mutex_store = PTHREAD_MUTEX_INITIALIZER;
+    mqtt_mutex = &mqtt_mutex_store;
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+
+    recv_msg_list = ListInitialize();
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if ((rc = pthread_mutex_init(mqtt_mutex, &attr)) != 0) {
+        printf("init: error %d initializing mqtt_mutex\n", rc);
+        MQTTHandle_error(rc);
+    } else {
+
+        conn_options.keepAliveInterval = 20;
+        conn_options.cleansession = 1;
+
+        conn_options.onSuccess = onConnect;
+        conn_options.onFailure = onConnectFailure;
+        conn_options.context = mqtt_client;
+
+        char clientid[9] = "no_id";
+        srand((unsigned int)time(0));
+        sprintf(clientid, "client_%d", rand()%100);
+
+        if ((rc = MQTTAsync_create(&mqtt_client, mqtt_server, clientid, MQTTCLIENT_PERSISTENCE_NONE, NULL)) != 0) {
+            printf("Failed to create Client, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+
+        MQTTAsync_setCallbacks(mqtt_client, NULL, NULL, mqtt_msg_arrived, NULL);
+        if ((rc = MQTTAsync_connect(mqtt_client, &conn_options)) != MQTTASYNC_SUCCESS) {
+            printf("Failed to start connect, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+        connection_status = MQTT_CONNECTED;
+    }
+}
+
+void AP_Telemetry_MQTT::set_mqtt_server(const char* server)
+{
+    mqtt_server = server;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_user(const char* user)
+{
+    conn_options.username = user;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_password(const char* password)
+{
+    conn_options.password = password;
+}
+
+void AP_Telemetry_MQTT::send_log(const char* str)
+{
+    char topic[MAX_PAYLOAD];
+    if (send_log_flag == MQTT_SEND_LOG_ON) {
+        if (connection_status == MQTT_CONNECTED) {
+            sprintf(topic, "$ardupilot/copter/quad/log/%04d/location",
```

Comment:
I think that the directory name general purpose is better. For example, "quad" thinks individual.